app/models/akka/SubscriptionChannel.scala

```diff
@@ -0,0 +1,107 @@
+package models.akka
+
+import akka.actor.{ActorRef, ActorSystem}
+import akka.stream._
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler, StageLogging}
+import models.rpc.{DSAMessage, ResponseMessage, SubscriptionNotificationMessage}
+import akka.pattern.ask
+import akka.util.Timeout
+import models.akka.Messages._
+
+import scala.concurrent.duration._
+import scala.collection.mutable
+import scala.concurrent.{Await, ExecutionContext}
+
+
+class SubscriptionChannel(val store: ActorRef)
+                         (implicit actorSystem: ActorSystem, materializer: Materializer)
+  extends GraphStage[FlowShape[SubscriptionNotificationMessage, DSAMessage]] {
+
+  type Sid = Int
+
+  val in = Inlet[SubscriptionNotificationMessage]("Subscriptions.in")
+  val out = Outlet[DSAMessage]("Subscriptions.out")
+  implicit val as = actorSystem
+  implicit val m = materializer
+  implicit val ctx = ExecutionContext.global
+
+  implicit val timeout = Timeout(3 seconds)
+
+  override def shape: FlowShape[SubscriptionNotificationMessage, DSAMessage] = FlowShape.of(in, out)
+
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with StageLogging {
+
+    def storeIt(message: SubscriptionNotificationMessage) = (store ? PutNotification(message)).mapTo[Int]
+
+
+    setHandler(in, new InHandler {
+      override def onPush(): Unit = {
+
+        log.debug(s"on push: ${in}")
+        val message = grab(in)
+
+        val callback = getAsyncCallback[Int]{
+          _=>
+            log.debug(s"storing $message")
```

Comment:
Interpolation may be costly here. Too bad Akka does not support trace(), but at least let's use `log.debug("...{}...{}...", arg1, arg2, ...)`. All in all, I'd probably remove some unnecessary logging from this class, I think this will declutter the output a bit.

Reply:
done